src/NativeScripts.purs

```diff
@@ -21,3 +41,66 @@ instance Show NativeScriptHash where
 nativeScriptHash :: NativeScript -> Maybe NativeScriptHash
 nativeScriptHash ns = wrap <<< Hashing.nativeScriptHash <$> convertNativeScript
   ns
+
+-- | `SetChoice` is an internal type representing internal state of
+-- | `getMaximumSigners` algorithm. There are a number of values (key hashes)
+-- | from which we must choose exactly `choice`. It's like a suspended `NOfK`
+-- | choice. We track the alternatives alongside their count to adjust for
+-- | possible duplicates in `allChoice`.
+type SetChoice a = Array { set :: Set a, choice :: Int }
+
+anyChoice :: forall a. Ord a => SetChoice a -> SetChoice a -> SetChoice a
+anyChoice as bs = Array.nub $ as <> bs
+
+allChoice :: forall a. Ord a => SetChoice a -> SetChoice a -> SetChoice a
+allChoice as bs = Array.concat do
+  a <- as
+  b <- bs
+  pure
+    -- We want to select a.choice from a.set and b.choice from b.set,
+    -- but there are equal elements that count for both if we select them.
+    -- So to satisfy the requirement we must uncount the elements present in
+    -- both sets.
+    [ { choice: a.choice + b.choice - Set.size (Set.intersection a.set b.set)
+      , set: a.set <> b.set
+      }
+    ]
```

Comment:
I'm not sure about performance, but `Array.concat` is just `Bind.join`, thus you can just drop `Array.concat` and the final `pure`

`join (m >>= f >>> pure) ≡ m >>= f`